### ISSUE TITLE
Sub-optimal results by overlap resolver

### ISSUE THREAD

**Issue by thorstenwagner** [thorstenwagner/ij-ridgedetection#15]
@hinerm 

I've played around with some other image data and found some strange results. The original image is:
http://fiji.sc/_images/f/f5/CNT_Ridge_Detection_Original.png

I've used the following settings...
Sigma = 1.5, Lower Threshold = 1.3, Higher Threshold = 7
... and get some 'strange' results:
![prob1](https://cloud.githubusercontent.com/assets/1113977/11099735/5bb8149c-88ad-11e5-86f3-7a5d426f9364.png)
![prob3](https://cloud.githubusercontent.com/assets/1113977/11099736/5d9a343e-88ad-11e5-9be2-ade187fa0907.png)
![prob2](https://cloud.githubusercontent.com/assets/1113977/11099746/62e004a0-88ad-11e5-85b1-4ee60506c005.png)

**Comment by hinerm** [comment]
thanks for playing with this @thorstenwagner!

So, for the bottom image (lines 133, 132) I think it is fairly arbitrary - from a slope/straightness perspective - which "top" end to pair with that "bottom" end pointing to the left.

My current implementation does a greedy search for maximizing straightness. We could improve the results by trying to find a globally optimal configuration for all pairings about an overlap point, which would improve the algorithm in some cases.

However, the primary assumption of a slope/straightness heuristic (that intersection is more likely than overlapping, simultaneous curves) falls apart the larger the length of overlap - especially when the length of overlap is large relative to the overall line length.

So I would say the 132, 133 line set is not an ideal candidate for the slope heuristic and would be better served by SVM, manual inspection, etc..

(and I will try to make this clear in my wiki contribution)

**Comment by hinerm** [comment]
Same explanation for the above two images. The initial segmentation is a poor starting point for a slope/straightness-based heuristic.

I played with the parameters a bit:

![screenshot from 2015-11-11 13 29 57](https://cloud.githubusercontent.com/assets/1123809/11100773/e0dfe44a-8878-11e5-9c0b-c0976c726ca3.png)
![screenshot from 2015-11-11 13 30 04](https://cloud.githubusercontent.com/assets/1123809/11100776/e22850d0-8878-11e5-8f02-743ac43d0436.png)

which looks opposite of what you might expect at first, but I think it's working exact as advertised. The algorithm likely paired the top two sections together because they were processed first (greedy) and they best preserve overall straightness. Then the bottom two sections are simply left over and paired together.

My goal for the wiki is to contrast these examples with images like this one:

![c1-soapes](https://cloud.githubusercontent.com/assets/1123809/11100879/670d1010-8879-11e5-93bf-85b0c9475ebe.png)

to show when slope overlap detection is, and is not, useful.

**Comment by hinerm** [comment]
There are, of course, probably still bugs! So let me know if you find more odd behavior :smile: 

Thanks again!

**Comment by hinerm** [comment]
![screenshot from 2015-11-11 14 51 14](https://cloud.githubusercontent.com/assets/1123809/11102720/e13549e8-8883-11e5-93f3-a633c2f29531.png)

Here is something that should never happen. The wrong ends were merged together.

**Comment by thorstenwagner** [comment]
@hinerm Yes, but it seems only happen sometimes... If check / uncheck preview multiple times it only occurs sometimes...

**Comment by hinerm** [comment]
I will investigate!

**Comment by hinerm** [comment]
So the problem is - unsurprisingly - that I took some short cuts in simplifying the assumptions I was making about intersection structure.

I [started a branch](https://github.com/hinerm/ij-ridgedetection/tree/improve-slope-search) to work on making the search more robust - by considering all combinations of "enclosed" segments as possible pairings, which then will necessitate global error minimization as well - but I don't really have the time to finish it right now.

It's fun to work on so I hope to get back to it eventually. Anyone else is welcome to continue its refinement as well. And @thorstenwagner if there's anything you consider a showstopper bug I will try to fix it.
